src/pymor/algorithms/projection.py

```diff
@@ -242,6 +243,23 @@ def action_BlockOperatorBase(self, op):
         else:
             return np.sum(projected_ops)
 
+    @match_class(BilinearFunctional)
+    def action_BilinearFunctional(self, op):
+        if op.operator.source == op.operator.range:
+            return BilinearFunctional(project(op.operator, self.source_basis, self.source_basis), name=op.name)
+        else:
+            return BilinearFunctional(project(op.operator, self.range_basis, self.source_basis), name=op.name)
+
+    @match_class(BilinearProductFunctional)
```

Comment:
This actually never gets called since `math_class(BilinearFunctional)` will also match `BilinearProductFunctional` and will be considered first.
Also, now that a `BilinearProductFunctional` is a `BilinearFunctional` one should ask oneself if a dedicated rule is needed or whether always getting a `BilinearFunctional` is fine.